stan/math/prim/mat/fun/add_diag.hpp

```diff
@@ -0,0 +1,87 @@
+#ifndef STAN_MATH_PRIM_MAT_FUN_ADD_DIAG_HPP
+#define STAN_MATH_PRIM_MAT_FUN_ADD_DIAG_HPP
+
+#include <stan/math/prim/mat/fun/Eigen.hpp>
+#include <stan/math/prim/scal/err/check_not_nan.hpp>
+#include <stan/math/prim/scal/err/check_finite.hpp>
+#include <stan/math/prim/scal/err/check_greater_or_equal.hpp>
+#include <stan/math/prim/scal/fun/fmin.hpp>
+#include <stan/math/prim/scal/meta/return_type.hpp>
+
+namespace stan {
+namespace math {
+/**
+ * Returns a Matrix with a constant added along the main diagonal.
+ *
+ * @tparam T_m type of element in Eigen::Matrix
+ * @tparam T_a type of element to add along the diagonal
+ *
+ * @param mat a matrix of dimensions greater than 1, 1
+ * @param to_add a constant to add along the diagonal
+ * @return a matrix with a constant added along main diagonal
+ */
+template <typename T_m, typename T_a>
+inline typename Eigen::Matrix<typename return_type<T_m, T_a>::type,
+                              Eigen::Dynamic, Eigen::Dynamic>
+add_diag(const Eigen::Matrix<T_m, -1, -1> &mat, const T_a &to_add) {
+  size_t length_diag;
+  length_diag = fmin(mat.rows(), mat.cols());
+  Eigen::Matrix<typename return_type<T_m, T_a>::type, -1, -1> out = mat;
```

Comment:
Putting a regular `Matrix` type on left, means the template expression on the right will assign.  That's very expensive in this case, because the left-hand side is a full matrix.  Instead, a loop can be used to increment.  
```
for (int k = 0; k < length_diag; ++k)
  out(k, k) += to_add;
```